Index scan support during UPDATE/DELETE on compressed hypertables.

sb230132 opened the pull request:
During UPDATE/DELETE on compressed hypertables, we do a sequential scan which can be improved by supporting index scans.

In this patch for a given UPDATE/DELETE query, if there are any WHERE conditions specified using SEGMENT BY columns, we use index scan to fetch all matching rows. Fetched rows will be decompressed and moved to uncompressed chunk and a regular UPDATE/DELETE is performed on the uncompressed chunk.

kgyrtkirk reviewed (commented):
didn't got to the bottom of it - but I think I've noted some things which might be usefull

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2061,32 +2123,25 @@ fill_predicate_context(Chunk *ch, List *predicates, List **segmentby_columns,
-						/* this constraint is not IS [NOT] NULL, so mark is_null_check as 0 */
-						*is_null_check = lappend_int(*is_null_check, 0);
+						/* Convert a qual clause to a form needed by the executor. */
+						linitial(opexpr->args) = rightop;
+						lsecond(opexpr->args) = leftop;
```
I have tested this on my patch, also i have testcase for <= AND >= on SEGMENTBY column 
```
test=# SELECT * FROM sample_table;
 time | device_id | val
------+-----------+-----
    1 |         1 |   1
    4 |         3 |
    3 |         4 |   1
    2 |           |   1
    6 |           |
   11 |         1 |   2
   13 |         4 |
   13 |         5 |   3
   10 |           |   2
   11 |           |   2
   12 |           |
(11 rows)

test=# BEGIN;
BEGIN
test=*# SELECT * FROM sample_table WHERE device_id > 4;
 time | device_id | val
------+-----------+-----
   13 |         5 |   3
(1 row)

test=*# UPDATE sample_table set time = 17 WHERE device_id > 4;
UPDATE 1
test=*# SELECT * FROM sample_table WHERE device_id > 4;
 time | device_id | val
------+-----------+-----
   17 |         5 |   3
(1 row)

test=*# rollback;
ROLLBACK

test=# begin;
BEGIN

test=*# SELECT * FROM sample_table WHERE device_id <=2;
 time | device_id | val
------+-----------+-----
    1 |         1 |   1
   11 |         1 |   2
(2 rows)

test=*# DELETE FROM sample_table WHERE device_id <=2;
DELETE 2
test=*# SELECT * FROM sample_table WHERE device_id <=2;
 time | device_id | val
------+-----------+-----
(0 rows)

test=*# ROLLBACK;
```

sb230132 replied:
Tested my patch with != condition and i face an error.

```
test=*# DELETE FROM sample_table WHERE device_id != 2;
ERROR:  operator 518 is not a member of opfamily 1976
```

Any idea how to fix it ? I think i need to call op_in_opfamily() to check i given operator belong to op family.

sb230132 replied:
Fixed and added few more testcases.

sb230132 replied:
Fixed.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2021,21 +2022,82 @@ is_segmentby_col(List *ht_compression_info, char *column_name)
+	AttrNumber attnum = get_attnum(RelationGetRelid(comp_chunk_rel), column_name);
+	for (i = 0; i < index_rel->rd_index->indnatts; i++)
+	{
+		if (index_rel->rd_index->indkey.values[i] == var->varattno ||
+			index_rel->rd_index->indkey.values[i] == attnum)
```
When physical layout of compressed and uncompressed chunk are different we will end up having different attribute number for a given SEGMENT BY column. Thus i had to have x == a || x == b

sb230132 replied:
Fixed.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2021,21 +2022,82 @@ is_segmentby_col(List *ht_compression_info, char *column_name)
+
+	key_columns = RelationGetIndexAttrBitmap(comp_chunk_rel, INDEX_ATTR_BITMAP_ALL);
+	index_oids = RelationGetIndexList(comp_chunk_rel);
+	index_oid = linitial_oid(index_oids);
```
That is what i noticed. For a compressed chunk if there is a SEGMENT BY column defined, then there will be only 1 index. I will add an assert in case there is no index (in case someone drops this index :))

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2021,21 +2022,82 @@ is_segmentby_col(List *ht_compression_info, char *column_name)
+	Relation comp_chunk_rel;
 	ListCell *lc;
+
+	comp_chunk = ts_chunk_get_by_id(ch->fd.compressed_chunk_id, true);
```
May be i can add a struct which holds opened table and index Relations and pass this struct to all functions where it is needed.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2340,42 +2484,52 @@ decompress_batches(Chunk *ch, ScanKeyData *scankeys, int num_scankeys, Bitmapset
-							   &segmentby_columns_value,
-							   &is_null_check,
-							   &is_null);
+		segmentby_predicates = get_segmentby_predicates(linitial(chunks), predicates);
```
This function get_segmentby_predicates() will parse the WHERE clause and return a list containing all SEGMENT BY columns specified in WHERE.
Calling same function for every chunk is a repetitive job, thus want to optimize it.

kgyrtkirk replied:
I think the case when multiple chunks are in action is most likely uncovered - and I would be more worried about that there are mismatches between the chunks because of column drops in-between chunk creations - than to try to optimize that case.

In fact in most cases 1 chunk will be affected.
As about optimizing this: the cost of opening a table scan I think is magnitudes higher than running this predicates processing function.
In case for some reason this method does lit up during performace profiling - we can optimize it that time...

I think an assert would be better; or place it inside the below loop ; so that it doesn't run into those issues...

sb230132 replied:
This is fixed as part of a different PR.

sb230132 commented on tsl/test/expected/compression_update_delete.out:
```diff
@@ -914,13 +984,13 @@ DELETE FROM sample_table WHERE device_id >= 4 AND val <= 1;
  count 
 -------
-     0
+     4
```
DELETE query ideally deletes 1 row which is coming from 1 of the chunks. Without this patch, all rows from compressed chunks are moved to uncompressed chunk. Now only required compressed rows are decompressed. Thus you see this change.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2083,21 +2084,90 @@ is_segmentby_col(List *ht_compression_info, char *column_name)
-fill_predicate_context(Chunk *ch, List *predicates, List **segmentby_columns,
-					   List **segmentby_columns_value, List **is_null_check, List **is_null)
+static List *
+get_segmentby_predicates(Chunk *ch, List *predicates)
```
Right, now i dont change any contents of passed in predicates.

svenklemm commented on tsl/test/sql/compression_update_delete.sql:
```diff
@@ -511,6 +511,39 @@ DELETE from sample_table WHERE device_id = 3 AND val IS NULL;
+SELECT * FROM sample_table WHERE 3 < device_id;
+ROLLBACK;
+
+-- test scalar expressions
+BEGIN;
+SELECT * FROM sample_table  WHERE (device_id) in (3, 30, 21);
+UPDATE sample_table SET time = time + 2 WHERE (device_id) in (3, 30, 21);
+SELECT * FROM sample_table  WHERE (device_id) in (3, 30, 21);
+ROLLBACK;
+
+-- test row comparisons
+BEGIN;
+SELECT * FROM sample_table  WHERE row(device_id, val) > row(4, 1);
+UPDATE sample_table SET time = 18 WHERE row(device_id, val) > row(4, 1);
+SELECT * FROM sample_table  WHERE row(device_id, val) > row(4, 1);
+ROLLBACK;
```
What are this tests supposed to verify? UPDATE with those expressions already worked before this feature was added. I would expect tests to verify that the query actually uses an index scan.

svenklemm replied:
This needs tests that actually show we are using index scans. None of the tests do that.
I agree with Zoltan that you should probably split this PR.

sb230132 replied:
Ok, removed code related to scalar and row comparison expressions.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2160,21 +2230,81 @@ fill_predicate_context(Chunk *ch, List *predicates, List **segmentby_columns,
+				}
+			}
+			break;
+			case T_ScalarArrayOpExpr:
```
Removed.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2160,21 +2230,81 @@ fill_predicate_context(Chunk *ch, List *predicates, List **segmentby_columns,
 				}
 			}
 			break;
+			case T_RowCompareExpr:
```
Removed.

antekresic commented on tsl/test/sql/compression_update_delete.sql:
```diff
@@ -511,6 +511,39 @@ DELETE from sample_table WHERE device_id = 3 AND val IS NULL;
 SELECT * FROM sample_table WHERE device_id = 3 AND val IS NULL;
 ROLLBACK;
 
+-- test in equality condition
```
These tests don't use any decompression optimizations? Can we say that explicitly here?

sb230132 replied:
Added comment.

antekresic commented on tsl/test/sql/compression_update_delete.sql:
```diff
@@ -941,3 +974,27 @@ SELECT COUNT(*) FROM :COMPRESS_CHUNK_1 WHERE c4 > 5 AND _ts_meta_min_2 <= 5 and
 -- report true
 SELECT COUNT(*) = :total_rows - :total_affected_rows FROM sample_table;
 ROLLBACK;
+
```
This test doesn't really confirm we are using the index, they should work without one in an identical way.

Maybe we can use some debug messaging to confirm we are actually using the index.

Would also like a test with and without index and confirm we are getting same results.

sb230132 replied:
Now i output some debug info. Check below:

```
UPDATE sample_table SET c3 = c3 + 0 WHERE c4 = 4 AND c1 >= 7;
LOG:  statement: UPDATE sample_table SET c3 = c3 + 0 WHERE c4 = 4 AND c1 >= 7;
DEBUG:  Number of compressed rows fetched from index: 10 
Number of compressed rows filtered by orderby columns 7
```

antekresic commented on tsl/src/compression/compression.c:
```diff
@@ -2376,12 +2642,29 @@ decompress_batches_for_update_delete(List *chunks, List *predicates)
+																	 scankeys,
+																	 num_scankeys,
+																	 &chunk_status_changed))
+			{
+				if (chunk_status_changed == true)
+					ts_chunk_set_partial(chunk);
+			}
```
This seems duplicated for index and no-index decompression. We should refactor to have a single place where we change the chunk status.

sb230132 replied:
Fixed.

antekresic reviewed (commented):
Lets add some more testing to confirm index is actually used.

antekresic reviewed (approved):
Approved with minor nit/question. :tada:

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2255,6 +2390,146 @@ build_update_delete_scankeys(RowDecompressor *decompressor, List *filters, int *
+									true,
+									&tmfd,
+									false);
+		if (result != TM_Ok)
```
Though we do an index scan, we are fetching the compressed tuple for compressed chunk and then inserting into uncompressed chunk, and then deleting the compressed tuple from compressed chunk.
So yes we need to drop the tuple.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2190,16 +2297,31 @@ fill_predicate_context(Chunk *ch, List *predicates, List **filters, List **is_nu
+						OpExpr *newclause = NULL;
+						Expr *newvar = NULL;
+						newclause = makeNode(OpExpr);
+						memcpy(newclause, (OpExpr *) node, sizeof(OpExpr));
```
Iam writing contents of an existing clause into newclause.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2255,6 +2390,146 @@ build_update_delete_scankeys(RowDecompressor *decompressor, List *filters, int *
+build_update_delete_index_scankeys(RowDecompressor *decompressor, List *predicates,
+								   int *num_scankeys, EState *estate)
+{
+	List *index_oids = RelationGetIndexList(decompressor->in_rel);
```
At user level its an UPDATE or DELETE operation.
To achieve this at planner level, one should tell postgresql to perform below steps for a given hypertable:
1. Check for compressed chunk and the indexes on SEGMENT BY columns.
2. Scan the index, fetch the matching tuples.
     1. Decompress the matched tuple.
     2. Insert decompressed tuples into uncompressed chunk.
     3. Delete the fetched compressed tuple from compressed chunk.
3. Do a regular UPDATE or DELETE operation on the uncompressed chunk.

Achieving above at the planner phase, iam not sure we can do it.

What you suggest as an alternative, which is about deciding which scan is better based on predicates, i think we are again reiterating the work of optimiser which generates different paths/plans based on statistics and then decides the cheapest plan.

This PR improves a bit more compared to existing design in which we are doing a sequential scan. Now we do an index scan (assuming there is always an index defined on SEGMENT BY columns).

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2340,42 +2484,52 @@ decompress_batches(Chunk *ch, ScanKeyData *scankeys, int num_scankeys, Bitmapset
+			List *index_oids;
+			index_oids = RelationGetIndexList(comp_chunk_rel);
+			Assert(index_oids != NULL);
+			Relation index_rel = index_open(linitial_oid(index_oids), AccessShareLock);
```
Fixed. Now i iterate over each index rather than assume that there is only 1 index on compressed chunk.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2255,6 +2390,146 @@ build_update_delete_scankeys(RowDecompressor *decompressor, List *filters, int *
+ *  4.delete this row from compressed chunk
+ */
+static bool
+decompress_batches_using_index(RowDecompressor *decompressor, ScanKeyData *index_scankeys,
```
As discussed made this function more generic and re-usable.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2061,6 +2063,69 @@ add_filter_column_strategy(char *column_name, StrategyNumber strategy, Const *va
+				/* skip if operator opno is not a member of opfamily */
+				if (opno && !op_in_opfamily(opno, opfamily))
+				{
+					index_close(index_rel, AccessShareLock);
```
that index will be getting closed twice

sb230132 replied:
Fixed.

svenklemm reviewed (changes requested):
Predicates for an index scan should be split into index conditions and generic filters and the generic filters should not repeat the predicates used as index conditions

sb230132 commented:
> Predicates for an index scan should be split into index conditions and generic filters and the generic filters should not repeat the predicates used as index conditions

I debugged below queries to check if iam processing index conditions and filters more than once or not.
I noticed that the conditions are placed in respective List to be processed only once.
ex: device_id > 3 is stored in index_filters List and "time" = '2000-01-06 05:30:00+05:30'::timestamp with time zone is stored in filters List of fill_predicate_context() for each chunk.
I think this confirms that the predicates are not duplicated and not processed more than once for each chunk.
```
test=*# explain DELETE FROM metrics_compressed WHERE device_id > 3 and time = '2000-01-06 05:30:00+05:30'::timestamp with time zone;                                                                                                                                                                              QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------------
 Custom Scan (HypertableModify)  (cost=0.28..8.30 rows=1 width=10)
   ->  Delete on metrics_compressed  (cost=0.28..8.30 rows=1 width=10)
         Delete on _hyper_5_10_chunk metrics_compressed_1
         ->  Index Scan using _hyper_5_10_chunk_metrics_compressed_time_idx on _hyper_5_10_chunk metrics_compressed_1  (cost=0.28..8.30 rows=1 width=10)
               Index Cond: ("time" = '2000-01-06 05:30:00+05:30'::timestamp with time zone)
               Filter: (device_id > 3)
(6 rows)

test=*# explain DELETE FROM metrics_compressed WHERE device_id > 3 and (time = '2000-01-06 05:30:00+05:30'::timestamp with time zone OR time =                                                                                             '2000-01-06 05:30:00+05:30'::timestamp with time zone);
                                                                            QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Custom Scan (HypertableModify)  (cost=0.25..8.27 rows=1 width=10)
   ->  Delete on metrics_compressed  (cost=0.25..8.27 rows=1 width=10)
         Delete on _hyper_5_10_chunk metrics_compressed_1
         ->  Index Scan using _hyper_5_10_chunk_metrics_compressed_device_id_time_idx on _hyper_5_10_chunk metrics_compressed_1  (cost=0.25..8.27 rows=1 width=10)
               Index Cond: ((device_id > 3) AND ("time" = '2000-01-06 05:30:00+05:30'::timestamp with time zone))
(5 rows)
```

svenklemm commented on CHANGELOG.md:
```diff
@@ -20,6 +20,7 @@ accidentally triggering the load of a previous DB version.**
 * #5584 Reduce decompression during constraint checking
 * #5530 Optimize compressed chunk resorting
 * #5639 Support sending telemetry event reports
+* #5586 Index scan support during UPDATE/DELETE
```
i think you need to rebase this and adjust how changelog entries are handled

sb230132 replied:
Fixed.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2539,65 +2666,158 @@ decompress_batches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num
 		}
-		row_decompressor_decompress_row(decompressor, NULL);
-		*chunk_status_changed = true;
+		elog(DEBUG1,
+			 "\nNumber of compressed rows fetched from index: %d \n"
+			 "Number of compressed rows filtered by orderby columns: %d",
```
This message format does not conform to postgres guidelines about error messages. Have a look at https://www.postgresql.org/docs/current/error-style-guide.html

sb230132 replied:
Its not error message. Its debug message to know how many rows were scanned from index and how many rows were filtered.
Please suggest alternative if message is not appropriate.

svenklemm commented on tsl/test/sql/CMakeLists.txt:
```diff
@@ -110,10 +110,10 @@ endif(CMAKE_BUILD_TYPE MATCHES Debug)
   if(CMAKE_BUILD_TYPE MATCHES Debug)
-    list(APPEND TEST_FILES chunk_utils_internal.sql)
+    list(APPEND TEST_FILES chunk_utils_internal.sql
+         compression_update_delete.sql)
   endif()
-  list(APPEND TEST_FILES compression.sql compression_update_delete.sql
-       compression_permissions.sql)
+  list(APPEND TEST_FILES compression.sql compression_permissions.sql)
```
why do we require a debug build for these tests now?

sb230132 replied:
Fixed.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2288,6 +2290,79 @@ add_filter_column_strategy(char *column_name, StrategyNumber strategy, Const *va
+		if (skip_this_index)
+		{
+			index_close(index_rel, AccessShareLock);
+			break;
```
There seems to be no tests added that test the index selection, especially when multiple indexes are present

sb230132 replied:
No matter what i try, i always end up with 1 index on compressed chunk. Please point me to a testcase where i can test this scenario.

Below case i have many indexes on uncompressed chunk, then only 1 index on corresponding compressed chunk.

```
test=# \d+ _timescaledb_internal._hyper_1_1_chunk
                                       Table "_timescaledb_internal._hyper_1_1_chunk"
  Column   |           Type           | Collation | Nullable | Default | Storage | Compression | Stats target | Description
-----------+--------------------------+-----------+----------+---------+---------+-------------+--------------+-------------
 time      | timestamp with time zone |           | not null |         | plain   |             |              |
 device_id | integer                  |           |          |         | plain   |             |              |
 v0        | integer                  |           |          |         | plain   |             |              |
 v1        | integer                  |           |          |         | plain   |             |              |
 v2        | double precision         |           |          |         | plain   |             |              |
 v3        | double precision         |           |          |         | plain   |             |              |
Indexes:
    "_hyper_1_1_chunk_metrics_compressed_device_id_time_idx" btree (device_id, "time")
    "_hyper_1_1_chunk_metrics_compressed_time_idx" btree ("time")
    "_hyper_1_1_chunk_metrics_compressed_v0_idx" btree (v0)
    "_hyper_1_1_chunk_metrics_compressed_v1_idx" btree (v1)
    "_hyper_1_1_chunk_metrics_compressed_v2_idx" btree (v2)
    "_hyper_1_1_chunk_metrics_compressed_v3_idx" btree (v3)
Check constraints:
    "constraint_1" CHECK ("time" >= '1999-12-30 05:30:00+05:30'::timestamp with time zone AND "time" < '2000-01-06 05:30:00+05:30'::timestamp with time zone)
Inherits: metrics_compressed
Access method: heap

test=# \d+ _timescaledb_internal.compress_hyper_2_6_chunk
                                                Table "_timescaledb_internal.compress_hyper_2_6_chunk"
        Column         |                 Type                  | Collation | Nullable | Default | Storage  | Compression | Stats target | Description
-----------------------+---------------------------------------+-----------+----------+---------+----------+-------------+--------------+-------------
 time                  | _timescaledb_internal.compressed_data |           |          |         | external |             | 0            |
 device_id             | integer                               |           |          |         | plain    |             | 1000         |
 v0                    | _timescaledb_internal.compressed_data |           |          |         | external |             | 0            |
 v1                    | _timescaledb_internal.compressed_data |           |          |         | external |             | 0            |
 v2                    | _timescaledb_internal.compressed_data |           |          |         | external |             | 0            |
 v3                    | _timescaledb_internal.compressed_data |           |          |         | external |             | 0            |
 _ts_meta_count        | integer                               |           |          |         | plain    |             | 1000         |
 _ts_meta_sequence_num | integer                               |           |          |         | plain    |             | 1000         |
 _ts_meta_min_1        | timestamp with time zone              |           |          |         | plain    |             | 1000         |
 _ts_meta_max_1        | timestamp with time zone              |           |          |         | plain    |             | 1000         |
Indexes:
    "compress_hyper_2_6_chunk__compressed_hypertable_2_device_id__ts" btree (device_id, _ts_meta_sequence_num)
Inherits: _timescaledb_internal._compressed_hypertable_2
Access method: heap
Options: toast_tuple_target=128
```

svenklemm commented on tsl/test/isolation/expected/compression_dml_iso.out:
```diff
@@ -187,7 +190,7 @@ step DEL1: BEGIN; DELETE from ts_device_table WHERE location = 200;
 step UPDrr: BEGIN TRANSACTION ISOLATION LEVEL REPEATABLE READ; UPDATE ts_device_table SET value = 4 WHERE location = 200; <waiting ...>
 step DELc: COMMIT;
 step UPDrr: <... completed>
-ERROR:  could not serialize access due to concurrent update
+ERROR:  unexpected tuple lock status
```
why do these messages change?

sb230132 replied:
I see that return status from table_tuple_delete() was handled similarly in many different parts of the code. So following the same.

Previously for every return status from table_tuple_delete() we were reporting a different error. Now instead we report 1 single error message.

svenklemm replied:
Having distinct error messages for different situation is extremely useful when debugging. Throwing the same generic error message makes it harder to debug because it blurs what is actually happening.

sb230132 replied:
Ok. Fixed.

svenklemm reviewed (changes requested):
There still seem to be some problems:
```
CREATE table t2(time timestamptz,device text,value float);
SELECT create_hypertable('t2','time');
ALTER TABLE t2 SET (timescaledb.compress,timescaledb.compress_segmentby='device');
INSERT INTO t2 SELECT '2020-01-01','d1',1;
INSERT INTO t2 SELECT '2020-01-01','d1',2;
SELECT compress_chunk(show_chunks('t2'));

DROP INDEX _timescaledb_internal._compressed_hypertable_4_device__ts_meta_sequence_num_idx;
CREATE INDEX ON _timescaledb_internal._compressed_hypertable_4 (_ts_meta_min_1);
CREATE INDEX ON _timescaledb_internal._compressed_hypertable_4 (device,_ts_meta_sequence_num);

sven@dev[937130]=# BEGIN;DELETE FROM t2 WHERE device='d1';ROLLBACK;
BEGIN
Time: 1.610 ms
ERROR:  XX000: operator 98 is not a member of opfamily 434
LOCATION:  get_op_opfamily_properties, lsyscache.c:147
Time: 2641.740 ms (00:02.642)
ROLLBACK
Time: 1.379 ms


```

sb230132 commented:
> There still seem to be some problems:
> 
> ```
> CREATE table t2(time timestamptz,device text,value float);
> SELECT create_hypertable('t2','time');
> ALTER TABLE t2 SET (timescaledb.compress,timescaledb.compress_segmentby='device');
> INSERT INTO t2 SELECT '2020-01-01','d1',1;
> INSERT INTO t2 SELECT '2020-01-01','d1',2;
> SELECT compress_chunk(show_chunks('t2'));
> 
> DROP INDEX _timescaledb_internal._compressed_hypertable_4_device__ts_meta_sequence_num_idx;
> CREATE INDEX ON _timescaledb_internal._compressed_hypertable_4 (_ts_meta_min_1);
> CREATE INDEX ON _timescaledb_internal._compressed_hypertable_4 (device,_ts_meta_sequence_num);
> 
> sven@dev[937130]=# BEGIN;DELETE FROM t2 WHERE device='d1';ROLLBACK;
> BEGIN
> Time: 1.610 ms
> ERROR:  XX000: operator 98 is not a member of opfamily 434
> LOCATION:  get_op_opfamily_properties, lsyscache.c:147
> Time: 2641.740 ms (00:02.642)
> ROLLBACK
> Time: 1.379 ms
> ```

Thanks for the testcase.

I introduced a new function find_matching_index(). This function will return a index which has all columns specified in the WHERE predicate. This matching index will be used to do a scan, filter out the rows based on scankeys etc.

Added testcase too.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2689,65 +2993,143 @@ decompress_batches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num
+	IndexScan *node = makeNode(IndexScan);
+	Plan *plan = &node->scan.plan;
+
+	plan->targetlist = NULL;
+	plan->qual = predicates;
+	plan->lefttree = NULL;
+	plan->righttree = NULL;
+	node->scan.scanrelid = in_rel->rd_id;
+	node->indexid = index_rel->rd_id;
+	node->indexqual = predicates;
+	node->indexqualorig = NULL;
+	node->indexorderby = NULL;
+	node->indexorderbyorig = NULL;
+	node->indexorderbyops = NULL;
+	node->indexorderdir = ForwardScanDirection;
```
makeNode uses palloc0 so explicit NULL initialization is redundant
```suggestion
	plan->qual = predicates;
	node->scan.scanrelid = in_rel->rd_id;
	node->indexid = index_rel->rd_id;
	node->indexqual = predicates;
	node->indexorderdir = ForwardScanDirection;
```

sb230132 replied:
Fixed.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2777,8 +3161,15 @@ decompress_batches_for_update_delete(Chunk *chunk, List *predicates)
 	Bitmapset *null_columns = NULL;
 	int num_scankeys = 0;
+	List *index_scankeys = NIL;
+	List *num_index_scankeys = NIL;
```
why is this a list?

sb230132 replied:
Fixed.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2689,65 +2993,143 @@ decompress_batches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num
+ * scans on compressed chunks.
+ */
+static List *
+build_index_scankeys(Relation in_rel, Relation index_rel, List *predicates, List **num_scankeys,
```
Why is num_scankeys a list?

sb230132 replied:
Fixed.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2790,20 +3181,53 @@ decompress_batches_for_update_delete(Chunk *chunk, List *predicates)
+		{
+			ListCell *num_idx_keys;
+			ListCell *scan_idx_keys;
+			forboth (num_idx_keys, num_index_scankeys, scan_idx_keys, index_scankeys)
```
when would this list have more than one enrtry?

sb230132 replied:
Fixed.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2790,20 +3181,53 @@ decompress_batches_for_update_delete(Chunk *chunk, List *predicates)
+											  ordered_index_filters,
+											  &num_index_scankeys,
+											  estate);
+		if (index_scankeys)
```
under which circumstance would index_scankeys be NULL

sb230132 replied:
Fixed.

sb230132 commented on tsl/src/compression/compression.c:
```diff
@@ -2689,65 +2993,143 @@ decompress_batches(RowDecompressor *decompressor, ScanKeyData *scankeys, int num
+						   NULL);
+
+	*num_scankeys = lappend_int(*num_scankeys, indexstate->iss_NumScanKeys);
+	scankeys = lappend(scankeys, indexstate->iss_ScanKeys);
```
Fixed.

svenklemm commented on tsl/src/compression/compression.c:
```diff
@@ -2438,6 +2440,248 @@ add_filter_column_strategy(char *column_name, StrategyNumber strategy, Const *va
+		{
+			Oid opfamily = index_rel->rd_opfamily[i];
+			/* assert if operator opno is not a member of opfamily */
+			if (opno && !op_in_opfamily(opno, opfamily))
+				Assert(false);
```
```suggestion
			Assert(opno && op_in_opfamily(opno, opfamily));
```

svenklemm replied:
What happens if opno is not in opfamily? Do we have a test for this?
